op-supervisor: Cross safe updates cycle detection + tests

tyler-smith opened the pull request:
Overview of changes:


- Fixed edge cases
  - Self-referencing messages
  - Unknown chains
  - Out-of-timestamp dependencies
  - First message handling
- Increased testing
  - Table-driven tests for better test vector management
  - Extensive cycle detection test scenarios
- Separated graph construction from cycle checking for easier testing and debugging 
- Collects data from each chain before construction so we can know when target nodes are out-of-timestamp
- The `HazardCycleChecks` routine now looks like:
  - Collect nodes from each chain
  - Build graph
  - Check graph for cycles using Kahn's algorithm

**metadata**

Fix https://github.com/ethereum-optimism/optimism/issues/12682

protolambda committed:
op-supervisor: experimental cross-safety, with hazard detection

tcrypt25519 committed:
tweak: Add some errors/error returns in backend/cross.

tcrypt25519 committed:
wip: Chain index <> ID mapping.

tcrypt25519 committed:
fix: Check parent instead of re-checking hazardBlock.

axelKingsley committed:
Remove Hazard Work

axelKingsley committed:
Write missing DB Bindings OpenBlock, LocallyDerivedFrom, CrossDerivedFrom

axelKingsley committed:
Configurable WorkFn for Workers

protolambda committed:
op-supervisor: move chain-index <> chain ID translation into dependency set, fix some interfaces

protolambda committed:
op-supervisor: update cross-safety worker routine

protolambda committed:
op-supervisor: update more error handling

protolambda committed:
op-supervisor: move errors to types package

protolambda committed:
op-supervisor: check CanExecuteAt and CanInitiateAt

protolambda committed:
op-supervisor: determine cross-safe candidate and L1 scope, and more fixes

protolambda committed:
todo L1 scope increment

protolambda committed:
op-supervisor: cross-safe L1 scope bump

protolambda committed:
op-supervisor: dependency set getter

protolambda committed:
op-supervisor: L1 scope increment fix

protolambda committed:
op-supervisor: fix cross-safe updates typing

tcrypt25519 committed:
Add cycle.go and initial unit tests.

protolambda committed:
op-node: signal L1 traversal of derivation to supervisor

protolambda committed:
op-supervisor: fromda fixes and tests

protolambda committed:
op-supervisor: fix OpenBlock, fix/add missing interface methods, hook up cross-safe worker routines

axelKingsley commented:
FYI, `go-lint` check is failing:

```
golangci-lint run -E goimports,sqlclosecheck,bodyclose,asciicheck,misspell,errorlint --timeout 5m -e "errors.As" -e "errors.Is" ./...
op-supervisor/supervisor/backend/cross/cycle.go:43:3: SA9003: empty branch (staticcheck)
                if err != nil {
                ^
make: *** [Makefile:23: lint-go] Error 1

Exited with code exit status 2
```

axelKingsley committed:
OpenBlock to return map[uint32]ExecutingMessage

axelKingsley committed:
Add Frontier Unit Tests

axelKingsley committed:
fix WithParent panic

protolambda committed:
op-node: register L1 traversal with op-supervisor

protolambda committed:
op-node,op-supervisor: add logging, temp work around for interop local-safe updates

tcrypt25519 committed:
fix: Use new execMsg log index info to complete graph.

tcrypt25519 committed:
debug: Add helper to write out Mermaid of our graph.

tcrypt25519 committed:
tests: Add more cycle tests.

axelKingsley committed:
Add safe_start_test, unsafe_start_test

axelKingsley committed:
Add safe_update_test and unsafe_update_test

axelKingsley committed:
add worker_test

protolambda committed:
op-supervisor: fix cross-safe L1 scope bumping

protolambda committed:
op-supervisor: fix logs DB test

tcrypt25519 committed:
tests: Organize cycle tests.

tcrypt25519 committed:
tests: Add NoCycle test.

tcrypt25519 committed:
fix: Add edges between basic logs.

tcrypt25519 committed:
tests: More comprehensive NoCycle tests.

tcrypt25519 committed:
fix: Disallow self-referencing messages.

tcrypt25519 committed:
tests: Make tests not use index-0 exec msgs for now.

tcrypt25519 committed:
fix: Safely handle unknown chains.

tcrypt25519 committed:
tests,fix: Remove unintended self-reference.

tcrypt25519 committed:
tests,fix: Remove unintended self-reference.

tcrypt25519 committed:
tests: Add test for the first log being exec.

tcrypt25519 committed:
refactor: Create ErrFailedToOpenBlock for error.

tcrypt25519 committed:
tests,refactor: Use table-driven tests to make it easier to iterate test vectors.

tcrypt25519 committed:
cleanup: Change some comments and make Mermaid debugging easier to disable.

tcrypt25519 committed:
refactor: Split graph build from checks; abstract adding edges.

This allows us to more easily test specifics of our cycle checking.

Errors can be in either part, and only being able to test the entirety
makes things a little difficult.

Abstracting edge creation helps eliminate bugs like 118aeb6b8.

tcrypt25519 committed:
tests: Add permuations of 2-cycles involving adjacency and first logs.

tcrypt25519 committed:
fix: When adding edges for execMsgs, remove node from inDegree0.

tcrypt25519 committed:
tests: Test execMsg to base log which is right after an execMsg to the first execMsg.

tcrypt25519 committed:
tests: Cycle across timestamps should not return an error.

tcrypt25519 committed:
fix: Don't consider cycles that depend on out-of-timestamp nodes.

We do this by first collecting every execMsg from all chains, filtering
out the out-of-timestamp nodes, and putting into a lookup map.

When creating edges, we check to see if the target node is
uninteresting.

axelKingsley commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,300 @@
+			}
+
+			if i == 0 {
+				// First log in block has no dependencies.=
```
```suggestion
				// First log in block has no dependencies
```

axelKingsley commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,300 @@
+				// First log in block has no dependencies.=
+				g.inDegree0[k] = struct{}{}
+			} else {
+				// Add edge: prev log <> current log
+				prevKey := msgKey{
+					chainIndex: hazardChainIndex,
+					logIndex:   i - 1,
+				}
+				g.addEdge(prevKey, k)
```
I'm trying to understand why we need to establish Edges between all messages linearly like this. I'm imagining that we have some web of connections representing Executing Messages, but then we also have this linked-list of edges representing every log in the block. 

But, non-executing messages don't establish a dependency, and therefore I would think don't need to be included in the graph?

tyler-smith replied:
I believe this came about during a call with @protolambda when we discussed this log-oriented graph instead of the block-oriented one outlined in the original spec, but I don't recall the exact reasoning for the edges between sequential logs instead of only between executing and initiating msgs.

Trying it without these edges seems to make the cycle finding break in some of our cases so I think we need to really understand this piece.

axelKingsley commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,300 @@
+		return err
+	}
+
+	logMermaidDiagram("Built graph", g)
```
This does not need to be logged on every check I would assume

axelKingsley commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,300 @@
+			initMsg := execMsgs[m.Chain][m.LogIdx]
+			if initMsg == nil {
+				continue
+			}
```
Wouldn't this be `initMsg, ok :=...` if you're checking for the presence of an item in a map?

tcrypt25519 committed:
fix: Remove debug mermaid logging.

tcrypt25519 committed:
cleanup: Remove stray text in comment.

tcrypt25519 committed:
tweak: Make presence check more idiomatic; remove possible nil map access.

tcrypt25519 committed:
docs: Improve comments in cycle detection code.

tcrypt25519 committed:
cleanup: Reorganize cycle detection code.

tcrypt25519 committed:
cleanup: Move public types to top.

tcrypt25519 committed:
cleanup: Change some names for consistency.

tcrypt25519 committed:
cleanup: Better comment in cycle detection.

tcrypt25519 committed:
docs: Explain graph representation.

tcrypt25519 committed:
tests: Re-organize HazardCycleChecks tests.

tcrypt25519 committed:
docs: Better naming and docs in cycle.go.

tcrypt25519 committed:
tests: Add large graph cycle detection tests.

tcrypt25519 committed:
tests: Add nil hazard map test.

tcrypt25519 committed:
refactor: Slightly tweak graph sort code to nest less; add comments.

protolambda commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,303 @@
+			if !ok {
+				continue
+			}
+			if _, ok := initChainMsgs[m.LogIdx]; !ok {
+				continue
+			}
```
This should return an error (wrapping the `types.ErrConflict`) instead. If the executing message points to a non-existent log then there's no point in resolving any remaining cycles, as the block will be invalid.

tyler-smith replied:
This check is actually just checking that the init message is within the right timestamp. I've added an additional check to ensure the init message exists.

protolambda commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,303 @@
+				logIndex:   execLogIdx,
+			}
+
+			// Disallow self-referencing messages
+			if initKey == execKey {
+				return nil, ErrExecMsgSelfReference
+			}
```
Update the comment that the executing message contains a hash of the initiating message. So a message cannot be self-referential. So this error case should be dead code. Returning an error, that wraps `ErrConflicts` (to invalidate the block) seems best to me.

protolambda commented on op-supervisor/supervisor/backend/cross/cycle_test.go:
```diff
@@ -0,0 +1,511 @@
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("2", 0),
+						1: execMsgWithTimestamp("2", 1, 99),
```
It's technically not valid to allow an executing message to reference an initiating message with a higher timestamp. But this is checked in a previous stage, outside of cycle-detection. A comment about that here might be helpful to avoid future confusion about this test case.

protolambda commented on op-supervisor/supervisor/backend/cross/cycle_test.go:
```diff
@@ -0,0 +1,511 @@
+		{
+			name: "2-cycle across chains",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("2", 1),
+					},
+				},
+				"2": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 1),
+					},
+				},
```
Maybe it makes more sense if the `key == 0`, to create a cycle? Otherwise the cycle collapses to the same kind of self-referential log-index case.
So that would be `0: execMsg("1", 1),`

protolambda commented on op-supervisor/supervisor/backend/cross/cycle_test.go:
```diff
@@ -0,0 +1,511 @@
+	runHazardCycleChecksTestCaseGroup(t, "NoCycle", tests)
+}
+
+func TestHazardCycleChecksCycle(t *testing.T) {
```
A test case with a 3rd chain may be interesting, to show a cycle is detected, even if one of the chains isn't involved in the cycle.

tcrypt25519 committed:
tweak: Make self referencial messages fail with an ErrConflict.

tcrypt25519 committed:
tests: Add tests for both older/younger timestamps, and document that younger init msgs shouldn't occur.

tcrypt25519 committed:
tests: Add tests for 2-cycle with 3 chains.

tcrypt25519 committed:
tweak: Check that an exec msg's init msg exists.

tcrypt25519 committed:
tests: tweak test to use index 0 instead of 1.

tcrypt25519 committed:
cleanup: Remove unused function.

tynes commented on op-supervisor/supervisor/backend/cross/cycle.go:
```diff
@@ -0,0 +1,302 @@
+
+// buildGraph constructs a dependency graph from the hazard blocks.
+func buildGraph(d CycleCheckDeps, inTimestamp uint64, hazards map[types.ChainIndex]types.BlockSeal) (*graph, error) {
+	g := &graph{
```
Could be useful to have a `newGraph` helper

protolambda committed:
op-e2e: fix interop mock backend action test

CI try 2

protolambda commented:
CI gave a "infrastructure failure" error, so I force pushed a commit with updated commit message to retrigger CI fully

protolambda committed:
Merge branch 'develop' into cross-safe-updates-cycle-tests

protolambda commented:
Merged in develop, and changed the target branch, to avoid a rebase of ~40+ small commits. Didn't want to squash them, but also needed the base branch. So now the old draft commits of the base branch that this was built on top of end up here, but it'll get squashed by GitHub and produce a sensible diff.
